In [1]:
%cd /kaggle/working
!rm -R /kaggle/working/model-merging
!rm -R /kaggle/working/models
!mkdir -p /kaggle/working/models
!git clone https://github.com/NiccoloSicilian/model-merging.git


/kaggle/working
rm: cannot remove '/kaggle/working/model-merging': No such file or directory
rm: cannot remove '/kaggle/working/models': No such file or directory
Cloning into 'model-merging'...
remote: Enumerating objects: 2604, done.
remote: Counting objects: 100% (118/118), done.
remote: Compressing objects: 100% (21/21), done.
remote: Total 2604 (delta 107), reused 96 (delta 96), pack-reused 2486 (from 3)
Receiving objects: 100% (2604/2604), 773.12 KiB | 6.84 MiB/s, done.
Resolving deltas: 100% (1741/1741), done.


In [2]:
# 1. Ensure we are in the repo folder
%cd /kaggle/working/model-merging

# 2. Force reinstall the package in editable mode explicitly using 'uv pip'
!uv pip install -e .

# 3. Verify the installation
# This should list 'model-merging' and its location
!uv pip list | grep model-merging

/kaggle/working/model-merging
Using Python 3.12.12 environment at: /usr
Resolved 284 packages in 3.68s                                       
Prepared 181 packages in 47.44s                                          
Uninstalled 87 packages in 7.76s
Installed 181 packages in 581ms                             
 - aiohttp==3.13.2
 + aiohttp==3.9.3
 + aiohttp-retry==2.8.3
 - aiosignal==1.4.0
 + aiosignal==1.3.1
 - altair==5.5.0
 + altair==5.3.0
 + amqp==5.2.0
 - anyio==4.12.0
 + anyio==4.3.0
 + anypy==0.0.6
 + appdirs==1.4.4
 - arrow==1.4.0
 + arrow==1.3.0
 + asyncssh==2.14.2
 - atpublic==5.1
 + atpublic==4.1.0
 - attrs==25.4.0
 + attrs==23.2.0
 - babel==2.17.0
 + babel==2.14.0
 + bandit==1.7.8
 - beautifulsoup4==4.13.5
 + beautifulsoup4==4.12.3
 + billiard==4.2.0
 - black==25.12.0
 + black==24.3.0
 + blessed==1.20.0
 - blinker==1.9.0
 + blinker==1.7.0
 - boto3==1.42.10
 + boto3==1.34.79
 - botocore==1.42.10
 + botocore==1.34.79
 - cachetools==5.5.2
 + cachetools==5.3.3
 + celery==5.3.6
 -

In [8]:

# 1. Uninstall broken packages
!pip uninstall -y wandb numpy

# 2. Install a numpy version compatible with everything (<2.0)
# 3. Reinstall a fresh version of wandb
!pip install "numpy<2.0" wandb

Found existing installation: wandb 0.16.6
Uninstalling wandb-0.16.6:
  Successfully uninstalled wandb-0.16.6
Found existing installation: numpy 1.26.4
Uninstalling numpy-1.26.4:
  Successfully uninstalled numpy-1.26.4
  Using cached numpy-1.26.4-cp312-cp312-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (61 kB)
  Using cached wandb-0.24.1-py3-none-manylinux_2_28_x86_64.whl.metadata (12 kB)
Using cached numpy-1.26.4-cp312-cp312-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (18.0 MB)
Using cached wandb-0.24.1-py3-none-manylinux_2_28_x86_64.whl (23.0 MB)
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
bigframes 2.26.0 requires google-cloud-bigquery-storage<3.0.0,>=2.30.0, which is not installed.
model-merging 0.0.0 requires wandb==0.16.6, but you have wandb 0.24.1 which is incompatible.
grain 0.2.15 requires protobuf>=5.28.3, but you have protobuf 4.25.3

In [3]:
%cd /kaggle/working/model-merging
!PYTHONPATH=./src MODELS_PATH=/kaggle/working/models uv run scripts/evaluate_multitask_merging.py \
    core.entity=null \
    train.logging.logger.entity=null \
    train.logging.logger.mode=disabled \
    train.logging.logger.log_model=False merger=dual "merger.optimal_alphas.ViT-B-16={8:1.2, 14:1.1, 20:1.2}" nn/encoder=b16 merger.svd_compress_factor=1.0 benchmark=N14

/kaggle/working/model-merging
Using CPython 3.11.7
Creating virtual environment at: .venv
░░░░░░░░░░░░░░░░░░░░ [0/285] Installing wheels...                               warning: Failed to hardlink files; falling back to full copy. This may lead to degraded performance.
         If the cache and target directories are on different filesystems, hardlinking may not be supported.
         If this is intentional, set `export UV_LINK_MODE=copy` or use `--link-mode=copy` to suppress this warning.
Installed 285 packages in 14.45s                            
[22:09:58] INFO     PyTorch version 2.8.0 available.                ]8;id=329303;file:///kaggle/working/model-merging/.venv/lib/python3.11/site-packages/datasets/config.py\config.py]8;;\:]8;id=468972;file:///kaggle/working/model-merging/.venv/lib/python3.11/site-packages/datasets/config.py#54\54]8;;\
/kaggle/working/model-merging/.venv/lib/python3.11/site-packages/lightning_utilities/core/imports.py:14: UserWarning: pkg_resources i

svd_compress_factor of the models = null (next set to 1, no compression)
MINE alpha 1.3 (compress factor 1/num_task):
**ViT-32 8 tasks**
| Dataset     | THEIRS Acc | THEIRS Loss | THEIRS Norm Acc |   MINE Acc | MINE Loss | MINE Norm Acc |
|:------------|-----------:|------------:|----------------:|-----------:|----------:|--------------:|
| SUN397      |     0.7028 |      1.1064 |          0.9386 |     0.7049 |    1.0770 |        0.9414 |
| Cars        |     0.7415 |      0.8322 |          0.9289 |     0.7474 |    0.7984 |        0.9364 |
| RESISC45    |     0.8581 |      0.5138 |          0.9004 |     0.8849 |    0.4150 |        0.9285 |
| EuroSAT     |     0.9578 |      0.1511 |          0.9649 |     0.9611 |    0.1323 |        0.9683 |
| SVHN        |     0.8926 |      0.3810 |          0.9233 |     0.8756 |    0.4214 |        0.9058 |
| GTSRB       |     0.8943 |      0.4043 |          0.9040 |     0.9349 |    0.2608 |        0.9450 |
| MNIST       |     0.9871 |      0.0816 |          0.9926 |     0.9857 |    0.0856 |        0.9912 |
| DTD         |     0.6606 |      1.2684 |          0.8449 |     0.6995 |    1.0936 |        0.8946 |
| **Average** | **0.8368** |           — |      **0.9247** | **0.8493** |         — |    **0.9389** |

MINE alpha 1.3 (compress factor 1/num_task):
**ViT-32 14 tasks**
| Dataset       | THEIRS Acc | THEIRS Loss | THEIRS Norm Acc |   MINE Acc | MINE Loss | MINE Norm Acc |
| ------------- | ---------: | ----------: | --------------: | ---------: | --------: | ------------: |
| SUN397        |     0.6898 |      1.0703 |          0.9212 |     0.6887 |    1.1076 |        0.9197 |
| Cars          |     0.6847 |      0.9641 |          0.8579 |     0.6990 |    0.9421 |        0.8758 |
| RESISC45      |     0.8284 |      0.5694 |          0.8693 |     0.8349 |    0.5671 |        0.8761 |
| EuroSAT       |     0.9022 |      0.2774 |          0.9090 |     0.9181 |    0.2344 |        0.9250 |
| SVHN          |     0.6684 |      1.0469 |          0.6915 |     0.7730 |    0.7187 |        0.7997 |
| GTSRB         |     0.7651 |      0.8257 |          0.7733 |     0.8673 |    0.4970 |        0.8767 |
| MNIST         |     0.9364 |      0.3433 |          0.9416 |     0.9725 |    0.1546 |        0.9779 |
| DTD           |     0.6436 |      1.2789 |          0.8231 |     0.6766 |    1.2038 |        0.8653 |
| Flowers102    |     0.8270 |      0.7043 |          0.9202 |     0.8213 |    0.6980 |        0.9139 |
| PCAM          |     0.7516 |      0.5281 |          0.8713 |     0.8058 |    0.4459 |        0.9342 |
| FER2013       |     0.6279 |      1.0661 |          0.8808 |     0.6535 |    0.9793 |        0.9167 |
| OxfordIIITPet |     0.9166 |      0.2633 |          0.9888 |     0.9076 |    0.3167 |        0.9791 |
| STL10         |     0.9800 |      0.0677 |          1.0046 |     0.9783 |    0.0824 |        1.0028 |
| CIFAR100      |     0.7513 |      0.9090 |          0.8640 |     0.7306 |    1.0281 |        0.8402 |
| **Average**   |     0.7838 |           — |      **0.8798** |     0.8091 |         — |    **0.9074** |


MINE alpha 1.2 (compress factor 1/num_task)
**ViT-32 20 tasks**

| Dataset       |    Acc | Loss   |   Norm Acc |
|:--------------|-------:|:-------|-----------:|
| SUN397        | 0.6767 | 1.1291 |     0.9038 |
| Cars          | 0.6418 | 1.1482 |     0.8041 |
| RESISC45      | 0.8013 | 0.6612 |     0.8408 |
| EuroSAT       | 0.8885 | 0.3190 |     0.8951 |
| SVHN          | 0.6891 | 0.9639 |     0.7128 |
| GTSRB         | 0.7797 | 0.7833 |     0.7881 |
| MNIST         | 0.9767 | 0.1600 |     0.9821 |
| DTD           | 0.6309 | 1.3314 |     0.8068 |
| Flowers102    | 0.8045 | 0.7583 |     0.8952 |
| PCAM          | 0.7749 | 0.5020 |     0.8984 |
| FER2013       | 0.6351 | 1.0488 |     0.8910 |
| OxfordIIITPet | 0.9022 | 0.3177 |     0.9732 |
| STL10         | 0.9731 | 0.1003 |     0.9976 |
| CIFAR100      | 0.7159 | 1.0814 |     0.8233 |
| CIFAR10       | 0.9466 | 0.1770 |     0.9706 |
| Food101       | 0.8300 | 0.6003 |     0.9647 |
| FashionMNIST  | 0.8090 | 0.5858 |     0.8592 |
| EMNIST        | 0.9808 | 0.1336 |     0.9862 |
| KMNIST        | 0.2941 | 2.1416 |     0.2997 |
| RenderedSST2  | 0.6930 | 0.5863 |     1.0210 |
| **avg** | **0.7722** | **—** | **0.8657** |


MINE alpha 1.3 (no compression):
| Dataset       |    Acc | Loss   |   Norm Acc |
|:--------------|-------:|:-------|-----------:|
| SUN397        | 0.7007 | 1.0766 |     0.9358 |
| Cars          | 0.7086 | 0.9156 |     0.8878 |
| RESISC45      | 0.7887 | 0.7211 |     0.8276 |
| EuroSAT       | 0.8933 | 0.2975 |     0.9    |
| SVHN          | 0.8051 | 0.6283 |     0.8329 |
| GTSRB         | 0.7939 | 0.7224 |     0.8025 |
| MNIST         | 0.9731 | 0.1418 |     0.9785 |
| DTD           | 0.6133 | 1.4148 |     0.7844 |
| Flowers102    | 0.7232 | 1.2193 |     0.8047 |
| PCAM          | 0.8188 | 0.4215 |     0.9492 |
| FER2013       | 0.6585 | 0.9767 |     0.9238 |
| OxfordIIITPet | 0.8921 | 0.3421 |     0.9624 |
| STL10         | 0.9697 | 0.1075 |     0.9941 |
| CIFAR100      | 0.7692 | 0.8778 |     0.8845 |
| avg           | 0.7935 | —      |     0.8906 |


MINE alpha 1.0 (no compression):
| Dataset | Acc | Loss | Norm Acc |
|--------|----:|-----:|---------:|
| SUN397 | 0.7057 | 1.0279 | 0.9425 |
| Cars | 0.7076 | 0.8965 | 0.8866 |
| RESISC45 | 0.7938 | 0.6678 | 0.8329 |
| EuroSAT | 0.8793 | 0.3229 | 0.8858 |
| SVHN | 0.7434 | 0.8150 | 0.7690 |
| GTSRB | 0.7317 | 0.9089 | 0.7396 |
| MNIST | 0.9534 | 0.2591 | 0.9587 |
| DTD | 0.6027 | 1.4361 | 0.7707 |
| Flowers102 | 0.7400 | 1.1358 | 0.8234 |
| PCAM | 0.8115 | 0.4597 | 0.9407 |
| FER2013 | 0.6495 | 1.0192 | 0.9111 |
| OxfordIIITPet | 0.8997 | 0.3138 | 0.9706 |
| STL10 | 0.9745 | 0.0899 | 0.9990 |
| CIFAR100 | 0.7762 | 0.8158 | 0.8926 |
| avg | 0.7835 | — | 0.8802 |


MINE alpha 1.2 (no compression):
| Dataset | Acc | Loss | Norm Acc |
|--------|----:|-----:|---------:|
| SUN397 | 0.7048 | 1.0528 | 0.9413 |
| Cars | 0.7094 | 0.9029 | 0.8888 |
| RESISC45 | 0.7921 | 0.6934 | 0.8311 |
| EuroSAT | 0.8911 | 0.3004 | 0.8978 |
| SVHN | 0.7881 | 0.6810 | 0.8153 |
| GTSRB | 0.7785 | 0.7743 | 0.7869 |
| MNIST | 0.9688 | 0.1711 | 0.9742 |
| DTD | 0.6144 | 1.4121 | 0.7857 |
| Flowers102 | 0.7313 | 1.1802 | 0.8138 |
| PCAM | 0.8215 | 0.4306 | 0.9523 |
| FER2013 | 0.6576 | 0.9854 | 0.9224 |
| OxfordIIITPet | 0.8992 | 0.3277 | 0.9700 |
| STL10 | 0.9721 | 0.1004 | 0.9965 |
| CIFAR100 | 0.7735 | 0.8493 | 0.8895 |
| avg | 0.7930 | — | 0.8904 |

MINE alpha 1.5 (no compression):
| Dataset | Acc | Loss | Norm Acc |
|--------|----:|-----:|---------:|
| SUN397 | 0.6907 | 1.1486 | 0.9224 |
| Cars | 0.7026 | 0.9599 | 0.8803 |
| RESISC45 | 0.7741 | 0.8048 | 0.8123 |
| EuroSAT | 0.8981 | 0.3041 | 0.9049 |
| SVHN | 0.8312 | 0.5462 | 0.8598 |
| GTSRB | 0.8164 | 0.6458 | 0.8252 |
| MNIST | 0.9806 | 0.1021 | 0.9860 |
| DTD | 0.6053 | 1.4496 | 0.7741 |
| Flowers102 | 0.6972 | 1.3343 | 0.7758 |
| PCAM | 0.8095 | 0.4154 | 0.9385 |
| FER2013 | 0.6601 | 0.9765 | 0.9259 |
| OxfordIIITPet | 0.8817 | 0.3854 | 0.9512 |
| STL10 | 0.9631 | 0.1268 | 0.9873 |
| CIFAR100 | 0.7604 | 0.9574 | 0.8744 |
| avg | 0.7908 | — | 0.8870 |

MINE alpha 1.4 (compression): 
| Dataset | Acc | Loss | Norm Acc |
|--------|----:|-----:|---------:|
| SUN397 | 0.6852 | 1.1309 | 0.9152 |
| Cars | 0.6976 | 0.9557 | 0.8739 |
| RESISC45 | 0.8275 | 0.5888 | 0.8683 |
| EuroSAT | 0.9219 | 0.2327 | 0.9287 |
| SVHN | 0.7905 | 0.6651 | 0.8178 |
| GTSRB | 0.8779 | 0.4489 | 0.8874 |
| MNIST | 0.9758 | 0.1288 | 0.9812 |
| DTD | 0.6766 | 1.2070 | 0.8653 |
| Flowers102 | 0.8118 | 0.7372 | 0.9034 |
| PCAM | 0.8102 | 0.4348 | 0.9393 |
| FER2013 | 0.6566 | 0.9694 | 0.9210 |
| OxfordIIITPet | 0.9035 | 0.3432 | 0.9747 |
| STL10 | 0.9762 | 0.0887 | 1.0008 |
| CIFAR100 | 0.7227 | 1.0731 | 0.8311 |
| avg | 0.8096 | — | 0.9077 |

MINE alpha 1.8 compression: 0.89

In [10]:

import ast
import re

raw_str = """
{'SUN397':               
                    [{'acc/test/SUN397':                                        
                    0.685239315032959,                                          
                    'loss/test/SUN397':                                         
                    1.1308931112289429,                                         
                    'normalized_acc/test/SUN39                                  
                    7': 0.9151584506034851}],                                   
                    'Cars': [{'acc/test/Cars':                                  
                    0.6975500583648682,                                         
                    'loss/test/Cars':                                           
                    0.9557216167449951,                                         
                    'normalized_acc/test/Cars'                                  
                    : 0.8739482760429382}],                                     
                    'RESISC45':                                                 
                    [{'acc/test/RESISC45':                                      
                    0.8274602890014648,                                         
                    'loss/test/RESISC45':                                       
                    0.5887748003005981,                                         
                    'normalized_acc/test/RESIS                                  
                    C45':                                                       
                    0.8682544827461243}],                                       
                    'EuroSAT':                                                  
                    [{'acc/test/EuroSAT':                                       
                    0.9218518733978271,                                         
                    'loss/test/EuroSAT':                                        
                    0.23271134495735168,                                        
                    'normalized_acc/test/EuroS                                  
                    AT': 0.928731381893158}],                                   
                    'SVHN': [{'acc/test/SVHN':                                  
                    0.790488600730896,                                          
                    'loss/test/SVHN':                                           
                    0.6651220321655273,                                         
                    'normalized_acc/test/SVHN'                                  
                    : 0.8177554607391357}],                                     
                    'GTSRB':                                                    
                    [{'acc/test/GTSRB':                                         
                    0.8779097199440002,                                         
                    'loss/test/GTSRB':                                          
                    0.4489012062549591,                                         
                    'normalized_acc/test/GTSRB                                  
                    ': 0.8873949646949768}],                                    
                    'MNIST':                                                    
                    [{'acc/test/MNIST':                                         
                    0.9757999777793884,                                         
                    'loss/test/MNIST':                                          
                    0.12883076071739197,                                        
                    'normalized_acc/test/MNIST                                  
                    ': 0.9811965823173523}],                                    
                    'DTD': [{'acc/test/DTD':                                    
                    0.6765957474708557,                                         
                    'loss/test/DTD':                                            
                    1.2070364952087402,                                         
                    'normalized_acc/test/DTD':                                  
                    0.8653061389923096}],                                       
                    'Flowers102':                                               
                    [{'acc/test/Flowers102':                                    
                    0.8118393421173096,                                         
                    'loss/test/Flowers102':                                     
                    0.7372411489486694,                                         
                    'normalized_acc/test/Flowe                                  
                    rs102':                                                     
                    0.9033659100532532}],                                       
                    'PCAM': [{'acc/test/PCAM':                                  
                    0.81024169921875,                                           
                    'loss/test/PCAM':                                           
                    0.43482011556625366,                                        
                    'normalized_acc/test/PCAM'                                  
                    : 0.9392910003662109}],                                     
                    'FER2013':                                                  
                    [{'acc/test/FER2013':                                       
                    0.6565895676612854,                                         
                    'loss/test/FER2013':                                        
                    0.9694204330444336,                                         
                    'normalized_acc/test/FER20                                  
                    13': 0.9210474491119385}],                                  
                    'OxfordIIITPet':                                            
                    [{'acc/test/OxfordIIITPet'                                  
                    : 0.9035159349441528,                                       
                    'loss/test/OxfordIIITPet':                                  
                    0.3432009518146515,                                         
                    'normalized_acc/test/Oxfor                                  
                    dIIITPet':                                                  
                    0.9747133255004883}],                                       
                    'STL10':                                                    
                    [{'acc/test/STL10':                                         
                    0.9762499928474426,                                         
                    'loss/test/STL10':                                          
                    0.0886918231844902,                                         
                    'normalized_acc/test/STL10                                  
                    ': 1.0007688999176025}],                                    
                    'CIFAR100':                                                 
                    [{'acc/test/CIFAR100':                                      
                    0.7226999998092651,                                         
                    'loss/test/CIFAR100':                                       
                    1.0731322765350342,                                         
                    'normalized_acc/test/CIFAR                                  
                    100':                                                       
                    0.8310717344284058}],                                       
                    'avg': [{'acc/test/avg':                                    
                    0.8095737227371761,                                         
                    'normalized_acc/test/avg':                                  
                    0.9077145755290985}]} 
"""
import ast
import re

# 1️⃣ Remove newlines ONLY (do NOT collapse spaces aggressively)
clean_str = raw_str.replace("\n", " ")

# 2️⃣ Parse safely
data = ast.literal_eval(clean_str)

def get_metric(metrics, prefix):
    """
    Return value whose key starts with prefix
    (ignores broken dataset names & spaces)
    """
    for k, v in metrics.items():
        if k.startswith(prefix):
            return v
    return None

# 3️⃣ Build table
rows = []
for dataset, values in data.items():
    metrics = values[0]

    acc  = get_metric(metrics, "acc/test/")
    loss = get_metric(metrics, "loss/test/")
    norm = get_metric(metrics, "normalized_acc/test/")

    rows.append((
        dataset,
        f"{acc:.4f}" if acc is not None else "—",
        f"{loss:.4f}" if loss is not None else "—",
        f"{norm:.4f}" if norm is not None else "—",
    ))

# 4️⃣ Print Kaggle-ready Markdown table
print("| Dataset | Acc | Loss | Norm Acc |")
print("|--------|----:|-----:|---------:|")
for r in rows:
    print(f"| {r[0]} | {r[1]} | {r[2]} | {r[3]} |")


| Dataset | Acc | Loss | Norm Acc |
|--------|----:|-----:|---------:|
| SUN397 | 0.6852 | 1.1309 | 0.9152 |
| Cars | 0.6976 | 0.9557 | 0.8739 |
| RESISC45 | 0.8275 | 0.5888 | 0.8683 |
| EuroSAT | 0.9219 | 0.2327 | 0.9287 |
| SVHN | 0.7905 | 0.6651 | 0.8178 |
| GTSRB | 0.8779 | 0.4489 | 0.8874 |
| MNIST | 0.9758 | 0.1288 | 0.9812 |
| DTD | 0.6766 | 1.2070 | 0.8653 |
| Flowers102 | 0.8118 | 0.7372 | 0.9034 |
| PCAM | 0.8102 | 0.4348 | 0.9393 |
| FER2013 | 0.6566 | 0.9694 | 0.9210 |
| OxfordIIITPet | 0.9035 | 0.3432 | 0.9747 |
| STL10 | 0.9762 | 0.0887 | 1.0008 |
| CIFAR100 | 0.7227 | 1.0731 | 0.8311 |
| avg | 0.8096 | — | 0.9077 |
